src/System.IO.Compression/tests/DeflateStreamAsyncTests.cs

```diff
@@ -0,0 +1,166 @@
+        {
+            byte[] buffer = new byte[32];
+            string testFilePath = Path.Combine("GZTestData", "Canterbury", "GZcompressed", "alice29.txt" + ".gz");
+            using (var readStream = await ManualSyncMemoryStream.GetStreamFromFileAsync(testFilePath, false, true))
+            using (var unzip = new DeflateStream(readStream, CompressionMode.Decompress, true))
+            {
+                Task task = null;
+                try
+                {
+                    task = unzip.ReadAsync(buffer, 0, 32);
+                    Assert.Throws<InvalidOperationException>(() => { unzip.ReadAsync(buffer, 0, 32); }); // "overlapping read"
+                }
+                finally
+                {
+                   // clean up
+                    readStream.SetManualEvent();
+                    // internal assert that this should complete 
+                    Assert.True(task.Wait(100));
+                }
+            }
+        }
+
+        [Fact]
+        public static async Task OverlappingFlushAsync()
+        {
+            byte[] buffer = new byte[32];
+            string testFilePath = Path.Combine("GZTestData", "Canterbury", "GZcompressed", "alice29.txt" + ".gz");
+            using (var readStream = await ManualSyncMemoryStream.GetStreamFromFileAsync(testFilePath, false, true))
+            using (var unzip = new DeflateStream(readStream, CompressionMode.Decompress, true))
+            {
+                Task task = null;
+                try
+                {
+                    task = unzip.ReadAsync(buffer, 0, 32);
+                    Assert.Throws<InvalidOperationException>(() => { unzip.FlushAsync(); }); // "overlapping flush"
+                }
+                finally
+                {
+                    // clean up
+                    readStream.SetManualEvent();
+                    // internal assert that this should complete 
+                    Assert.True(task.Wait(100));
+                }
+            }
+        }
+
+        [Fact]
+        public static async Task OverlappingWriteAsync()
+        {
+            byte[] buffer = null;
+            string testFilePath = Path.Combine("GZTestData", "Canterbury", "alice29.txt");
+            using (var origStream = await LocalMemoryStream.readAppFileAsync(testFilePath))
+            {
+                buffer = origStream.ToArray();
+            }
+
+            using (var writeStream = new ManualSyncMemoryStream(false))
+            using (var zip = new DeflateStream(writeStream, CompressionMode.Compress))
+            {
+                Task task = null;
+                writeStream.ResetManualEvent();
+                try
+                {
+                    task = zip.WriteAsync(buffer, 0, 32);
+                    // TODO: following code doesn't work sometimes
+                    // Assert.Throws<InvalidOperationException>(() => { zip.WriteAsync(buffer, 32, 32); }, "overlapping write");
```

Comment:
I'd like to use the Assert.Throws but I don't really understand why it fails and the regular try/catch doesn't. I tried several permutations of Assert.Throws and couldn't get it to pass.